### PR TITLE
CREATE VIEW fails in UNION if column datatypes are from different categories during MVU

### PR DESCRIPTION
**Pull request by rishabhtanwar29**
T-SQL has certain different rules and hooks for views
compared to PG and these rules and hooks are only
applicable in T-SQL dialect, that's why MVU fails while
restoring T-SQL views since restore happens in PG dialect.

Fix this by allowing T-SQL specific rules and hooks if
babelfishpg_tsql.dump_restore GUC is set to on.

Task: BABEL-3221
Signed-off-by: Rishabh Tanwar <ritanwar@amazon.com>

### CODE REVIEW

File: src/backend/parser/parse_coerce.c [hunk 1]
```diff
@@ -1380,6 +1381,7 @@ select_common_type(ParseState *pstate, List *exprs, const char *context,
 	TYPCATEGORY pcategory;
 	bool		pispreferred;
 	ListCell   *lc;
+	const char *dump_restore = GetConfigOption("babelfishpg_tsql.dump_restore", true, false);
```
Comment (surendravishnoi) [inline src/backend/parser/parse_coerce.c:1384]: I think we are using it at lot of places, may be better to have a backing global variable and use that for check. Also, it should be boolean GUC so values like ON/OFF TRUE/FALSE can all work

File: src/backend/parser/parse_coerce.c [hunk 2]
```diff
@@ -1437,8 +1439,9 @@ select_common_type(ParseState *pstate, List *exprs, const char *context,
 				pcategory = ncategory;
 				pispreferred = nispreferred;
 			}
-			else if (ncategory != pcategory
-				&& sql_dialect != SQL_DIALECT_TSQL) /* T-SQL allows to select common datatype between different categories */
+			else if (ncategory != pcategory &&
+				sql_dialect != SQL_DIALECT_TSQL && /* T-SQL allows to select common datatype between different categories */
+				(!dump_restore || (dump_restore && strcmp(dump_restore, "on") != 0))) /* allow common datatype between different categories if dump_restore GUC is set */
```
Comment (surendravishnoi) [inline src/backend/parser/parse_coerce.c]: comment should say "when restoring TSQL database", also dump_restore alone does not carry TSQL context, so we should alway specify that explicitly

File: src/backend/parser/parse_coerce.c (hunk 1 above)
Comment (rishabhtanwar29) [reply src/backend/parser/parse_coerce.c:1384]: Actually the guc is defined in tsql extension so we always have to fetch it's value using `GetConfigOption` function because if tsql extension is not installed then the guc will not even be defined. `extern` global variable will create engine's dependency on extension.

File: src/backend/parser/parse_coerce.c (hunk 2 above)
Comment (rishabhtanwar29) [reply src/backend/parser/parse_coerce.c]: Yup, updated the comment!

File: src/backend/parser/parse_coerce.c (hunk 1 above)
Comment (surendravishnoi) [reply src/backend/parser/parse_coerce.c:1384]: not for this change, 
most of the time options be used in engine only. May be it makes sense to move it to engine rather than keeping it in the extension.

**Review by surendravishnoi — APPROVED** [review]

File: src/backend/parser/parse_coerce.c [hunk 3]
```diff
@@ -1437,8 +1439,9 @@ select_common_type(ParseState *pstate, List *exprs, const char *context,
 				pcategory = ncategory;
 				pispreferred = nispreferred;
 			}
-			else if (ncategory != pcategory
-				&& sql_dialect != SQL_DIALECT_TSQL) /* T-SQL allows to select common datatype between different categories */
+			else if (ncategory != pcategory &&
+				sql_dialect != SQL_DIALECT_TSQL && /* T-SQL allows to select common datatype between different categories */
+				(!dump_restore || (dump_restore && strcmp(dump_restore, "on") != 0))) /* allow common datatype between different categories while restoring babelfish database */
```
Comment (2jungkook) [inline src/backend/parser/parse_coerce.c:1444]: Besides this case, am just wondering if there are any test cases for the additions below.

**Review by 2jungkook — APPROVED** [review]

File: src/backend/parser/parse_coerce.c (hunk 3 above)
Comment (rishabhtanwar29) [reply src/backend/parser/parse_coerce.c:1444]: Actually all the changes as a whole fixes the issue completely. The above condition will just skip the error when there are datatypes from different categories but in the `if` condition just below this, we will call `determine_datatype_precedence_hook` and `can_coerce_type` functions, `can_coerce_type` will eventually call `find_coercion_pathway_hook` hook to search into T-SQL coercion rules. But I will anyway add few more similar test cases to ensure we don't miss any edge case.
